algebra/vector_space.lean

```diff
@@ -0,0 +1,305 @@
+import algebra.ring
+import algebra.module
+
+universes u' v'
+
+class vector_space (α : inout Type u')  (β : Type v') [field α] extends module α β
+
+instance field.to_vector_space {K : Type} [field K] :
+vector_space K K := ⟨ring.to_module⟩
+
+lemma eq_zero_of_add_self_eq {α : Type} [add_group α]
+{a : α} (H : a + a = a) : a = 0 :=
+add_left_cancel (by {rw add_zero, exact H})
+
+class subspace (K : Type) (V : Type) [field K] [vector_space K V] (p : V → Prop) :=
+    (p_zero : p 0)
+    (p_add : ∀ u v, p u → p v → p (u + v))
+    (p_neg : ∀ v, p v → p (-v))
+    (p_smul : ∀ c v, p v → p (c • v))
```

Comment:
Drop the `p_` prefix from these, it doesn't make sense if `p` alpha-varies. Just refer to them with the prefix, i.e. `subspace.zero` etc.

Reply:
`p_neg` is not necessary, it can be proved using `p_smul`.

@digama0 @kckennylau what do you think of using a set `p : set V` instead of `p : V -> Prop`? It fells more natural to use sets.

Reply:
Also why is `subspace` a class?

Do we want to store `p` in itself? Then we can put a complete lattice structure on it?!
Otherwise I would prefer to explicitly put it into `Prop`, i.e. `structure subspace ... : Prop := ...`

Reply:
I suggested moving `p` out, he originally had it in a structure. But we could store it inside and then have a coercion for the projection.